ENIGMAsystem/SHELL/Collision_Systems/Polygon/Polygonimpl.cpp

```diff
@@ -0,0 +1,728 @@
+{
+    return collide_inst_ellipse(object, solid_only, prec, notme, x1, y1, r, r);
+}
+
+static bool line_ellipse_intersects(double rx, double ry, double x, double ly1, double ly2)
+{
+    // Formula: x^2/a^2 + y^2/b^2 = 1   <=>   y = +/- sqrt(b^2*(1 - x^2/a^2))
+    const double inner = ry*ry*(1 - x*x/(rx*rx));
+    if (inner < 0) {
+        return false;
+    }
+    else {
+        const double y1 = -sqrt(inner), y2 = sqrt(inner);
+        return y1 <= ly2 && ly1 <= y2;
+    }
+}
+
+enigma::object_collisions* const collide_inst_ellipse(int object, bool solid_only, bool prec, bool notme, int x1, int y1, double rx, double ry)
+{
+    // If wrong arguments return
+    if (rx == 0 || ry == 0)
+        return 0;
+
+    // Debugging Part Starts
+    // printf("Ellipse = (x1 = %d, y1 = %d, rx = %f, ry = %f)\n", x1, y1, rx, ry);
+    // Debugging Part Ends
+
+    // Iterate over the instances for the collision check
+    for (enigma::iterator it = enigma::fetch_inst_iter_by_int(object); it; ++it)
+    {
+        // Retrieving the instance
+        enigma::object_collisions* const inst = (enigma::object_collisions*)*it;
+
+        // Preliminary checks
+        if (notme && inst->id == enigma::instance_event_iterator->inst->id)
+            continue;
+        if (solid_only && !inst->solid)
+            continue;
+        if (inst->sprite_index == -1 && inst->mask_index == -1 && inst->polygon_index == -1) // No sprite/mask/polygon then no collision.
+            continue;
+
+        // Getting Bounding box of the instance
+        int left, top, right, bottom;
+        get_bbox_border(left, top, right, bottom, inst);
+
+        // Checking for bounding box collision
+        const bool intersects = line_ellipse_intersects(rx, ry, left-x1, top-y1, bottom-y1) ||
+                                 line_ellipse_intersects(rx, ry, right-x1, top-y1, bottom-y1) ||
+                                 line_ellipse_intersects(ry, rx, top-y1, left-x1, right-x1) ||
+                                 line_ellipse_intersects(ry, rx, bottom-y1, left-x1, right-x1) ||
+                                 (x1 >= left && x1 <= right && y1 >= top && y1 <= bottom); // Ellipse inside bbox.
+
+        // If the bbox collision is happening, we check for polygon only when
+        // prec is True AND there is a polygon for the instance
+        if (intersects) {
+            if (!prec) {
+                return inst;
+            }
+            if (inst->polygon_index == -1) {
+                return inst;
+            }
+            
+            // Actual polygon collision detection
+            // Retrieving the polygon
+            enigma::Polygon polygon1 = enigma::polygons.get(inst->polygon_index);
+
+            // Collision Detection
+            return get_polygon_ellipse_collision(inst->x, inst->y, x1, y1, polygon1, rx, ry)? inst : NULL;
+        }
+    }
+    return NULL;
+}
+
+void destroy_inst_point(int object, bool solid_only, int x1, int y1)
```

Comment:
These two routines are another great example: the collision system should provide a means of querying for instances at a point, and then some main function should use that collection or iterator to do the change logic. Otherwise we just end up with a bunch of identical implementations save for one line of effective change.

Reply:
I believe this comment is not yet resolved.

Reply:
I am sorry I couldn't understand this earlier and I thought It was resolved when I implemented this method. I agree this approach is very inefficient, and a cleaner approach would be to make a separate function for querying. I'll refactor this in an upcoming commit.

Reply:
This has been resolved in commit #22e732b. I am marking this as resolved